### PR TITLE
Initial specific access role changes

### PR DESCRIPTION
**Pull request by connorpgpmcelroy**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/EUI-5505, https://tools.hmcts.net/jira/browse/EUI-4326

### Change description ###
Adding approving specific request to node layer and connecting with real API

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x ] No
```

### CODE REVIEW

File: src/role-access/containers/specific-access/specific-access-duration/specific-access-duration.component.ts [hunk 1]
```diff
@@ -123,9 +131,30 @@ export class SpecificAccessDurationComponent implements OnInit {
...
+      requestId: 'ab9570d0-755e-4503-9af9-44b65c6f1717',
+      jurisdiction: 'IA',
+      roleCategory: RoleCategory.CASEWORKER,
+      requestedRole: 'specific-access-legal-ops',
+      person: {id: 'db17f6f7-1abf-4223-8b5e-1eece04ee5d8', name: null, domain: null}
+      }
       switch (navEvent) {
         case SpecificAccessNavigationEvent.CONTINUE:
-          this.store.dispatch(new fromFeature.ChangeSpecificAccessNavigation(SpecificAccessState.SPECIFIC_ACCESS_APPROVED));
+          /* TODO: at some point will need all four steps on node layer
```
Comment (udaydenduluri33) [inline src/role-access/containers/specific-access/specific-access-duration/specific-access-duration.component.ts:150]: can we remove this comment?

File: api/user/index.ts [hunk 2]
```diff
@@ -54,6 +54,20 @@ export async function refreshRoleAssignmentForUser(userInfo: UserInfo, req: any)
   return locationInfo;
 }
 
+/* export async function getRoleAssignmentForUser(userId: string, req: any): Promise<any []> {
```
Comment (udaydenduluri33) [inline api/user/index.ts:57]: Also delete this comment

File: api/accessManagement/index.ts [hunk 3]
```diff
@@ -33,6 +36,85 @@ export async function refreshRoleAssignments(req, res: Response, next: NextFunct
...
+  catch (error) {
+    next(error);
+    return res.status(error.status).send(error);
+  }
+}
+
+// attempts to delete 
+export async function deleteSpecificAccessRoles(req, res: Response, next: NextFunction, previousResponse: AxiosResponse<any>): Promise<Response> {
+  try {
+    req.body.requestId = req.body.rolesToDelete[1].id;
```
Comment (udaydenduluri33) [inline api/accessManagement/index.ts:81]: You do not need to set the req.body.requestId. You can pass it a parameter for the function deleteRoleByAssignmentId

File: api/accessManagement/index.ts [hunk 4]
```diff
@@ -33,6 +36,85 @@ export async function refreshRoleAssignments(req, res: Response, next: NextFunct
...
+    const specificAccessDeletionResponse = await deleteRoleByAssignmentId(req, res, next);
+    if (!specificAccessDeletionResponse || specificAccessDeletionResponse.status !== 204) {
+      // retry x 3
+      return previousResponse && previousResponse.status ? res.status(previousResponse.status).send(previousResponse) : res.status(400);
+    }
+    req.body.requestId = req.body.rolesToDelete[0].id;
```
Comment (udaydenduluri33) [inline api/accessManagement/index.ts:87]: You do not need to set the req.body.requestId. You can pass it a parameter for the function deleteRoleByAssignmentId

File: api/accessManagement/index.ts [hunk 5]
```diff
@@ -33,6 +36,83 @@ export async function refreshRoleAssignments(req, res: Response, next: NextFunct
...
+    }
+    req.body.restoreRole = caseRoles.find(role => role.id === req.body.requestId); */
+    // create the specific access approval role
+    const firstRoleResponse: AxiosResponse = await createSpecificAccessApprovalRole(req, res, next);
+    // 201
+    if (!firstRoleResponse || firstRoleResponse.status !== 201) {
+      return firstRoleResponse && firstRoleResponse.status ? res.status(firstRoleResponse.status).send(firstRoleResponse) : res.status(400);
+    }
+    const deletionResponse = await deleteRoleByAssignmentId(req, res, next, req.body.requestId);
+    req.body.rolesToDelete = firstRoleResponse.data.roleAssignmentResponse.requestedRoles;
```
Comment (udaydenduluri33) [inline api/accessManagement/index.ts:55]: `const rolesToDelete = firstRoleResponse.data.roleAssignmentResponse.requestedRoles;`

File: api/accessManagement/index.ts [hunk 6]
```diff
@@ -33,6 +36,84 @@ export async function refreshRoleAssignments(req, res: Response, next: NextFunct
...
+    return res.send(taskResponse.data).status(taskResponse.status);
+  } catch (error) {
+    next(error);
+    return res.status(error.status).send(error);
+  }
+}
+
+// attempts to delete
+// tslint:disable-next-line:max-line-length
+export async function deleteSpecificAccessRoles(req, res: Response, next: NextFunction, previousResponse: AxiosResponse<any>, rolesToDelete: any): Promise<Response> {
```
Comment (udaydenduluri33) [inline api/accessManagement/index.ts:77]: `rolesToDelete`  Can this be strongly typed?

Comment (connorpgpmcelroy) [reply api/accessManagement/index.ts:77]: yes, I'll do that

File: api/accessManagement/index.ts [hunk 7]
```diff
@@ -33,6 +36,84 @@ export async function refreshRoleAssignments(req, res: Response, next: NextFunct
...
+     ? res.status(previousResponse.status).send(previousResponse) : res.status(400);
+  } catch (error) {
+    next(error);
+    return res.status(error.status).send(error);
+  }
+}
+
+// attempts to restore the deleted specific access requested role on task completion failure
+// tslint:disable-next-line:max-line-length
+export async function restoreDeletedRole(req, res: Response, next: NextFunction, previousResponse: AxiosResponse<any>, rolesToDelete: any): Promise<Response> {
```
Comment (udaydenduluri33) [inline api/accessManagement/index.ts:102]: And here too `restoreDeletedRole` strongly typed

**Review by udaydenduluri33 — APPROVED** [review]
@connorpgpmcelroy  thanks for making the changes

File: api/accessManagement/index.ts [hunk 8]
```diff
@@ -33,6 +37,84 @@ export async function refreshRoleAssignments(req, res: Response, next: NextFunct
   return res.send(refreshRoleAssignmentsSuccess);
 }
 
+// node layer logic for approving specific access request
+export async function approveSpecificAccessRequest(req, res: Response, next: NextFunction): Promise<Response> {
+  try {
+    // attempted to find specific requested role previously to get data but not neccessary
+    /* const caseRoles = await getRoleAssignmentForUser(req.body.person.id, req);
+    if (!caseRoles) {
```
Comment (tarikdogan-hmcts) [inline api/accessManagement/index.ts:45]: can this commented code be removed

**Review by tarikdogan-hmcts — APPROVED** [review]

**Pull request merged**